R/unique-keys.R

```diff
@@ -0,0 +1,279 @@
+#' If `check == TRUE`, then it will first check if
+#' the given combination of columns is a unique key of the table.
+#'
+#' @inheritParams dm_add_pk
+#'
+#' @details The difference between a primary key (PK) and a unique key (UK) consists in the following:
+#' - when a local `dm` is copied to a database (DB) with `copy_dm_to()`, a PK will be set on the DB by default
+#' - a PK can be set as an `autoincrement` key (also implemented on certain DBMS when the `dm` is transferred to the DB)
+#' - there can be only one PK for each table, whereas there can be unlimited UKs
+#' - a UK will be used, if the same table has an autoincrement PK in addition, to ensure that during delta load processes
+#'   on the DB (cf. [dm_rows_append()]) the foreign keys are updated accordingly.
+#'   If no UK is available, the insertion is done row-wise, which also ensures a correct matching, but can be much slower.
+#' - a UK can generally enhance the data model by adding additional information
+#' - if a foreign key is added to point at a table without a corresponding PK or UK, a UK is automatically added to that table.
+#'
+#' @family primary key functions
+#'
+#' @return An updated `dm` with an additional unqiue key.
+#'
+#' @examplesIf rlang::is_installed("nycflights13") && rlang::is_installed("DiagrammeR")
+#'
+#' nycflights_dm <- dm(
+#'   planes = nycflights13::planes,
+#'   airports = nycflights13::airports,
+#'   weather = nycflights13::weather
+#' )
+#'
+#' # Create unique keys:
+#' nycflights_dm %>%
+#'   dm_add_uk(planes, tailnum) %>%
+#'   dm_add_uk(airports, faa, check = TRUE) %>%
+#'   dm_add_uk(weather, c(origin, time_hour)) %>%
+#'   dm_get_all_uks()
+#'
+#' # Keys can be checked during creation:
+#' try(
+#'   nycflights_dm %>%
+#'     dm_add_uk(planes, manufacturer, check = TRUE)
+#' )
+#' @export
+dm_add_uk <- function(dm, table, columns, ..., check = FALSE) {
+  check_dots_empty()
+
+  check_not_zoomed(dm)
+
+  table_name <- dm_tbl_name(dm, {{ table }})
+  table <- dm_get_tables_impl(dm)[[table_name]]
+
+  check_required(columns)
+  col_expr <- enexpr(columns)
+  col_name <- names(eval_select_indices(col_expr, colnames(table)))
+
+  if (check) {
+    table_from_dm <- dm_get_filtered_table(dm, table_name)
+    eval_tidy(expr(check_key(!!sym(table_name), !!col_expr)), list2(!!table_name := table_from_dm))
+  }
+
+  dm_add_uk_impl(dm, table_name, col_name)
+}
+
+# both "table" and "column" must be characters
+# in {datamodelr}, a primary key may consist of more than one columns
+# a key will be added, regardless of whether it is a unique key or not; not to be exported
+dm_add_uk_impl <- function(dm, table, column) {
+  def <- dm_get_def(dm)
+  i <- which(def$table == table)
+
+  if (!is_empty(def$pks[[i]]$column) && vctrs::vec_in(def$pks[[i]]$column[1], list(column))) {
+    abort_no_uk_if_pk(table, column)
+  }
+
+
+  def$uks[[i]] <- vctrs::vec_rbind(
+    def$uks[[i]],
+    new_uk(list(column))
+  )
```

Comment:
Should we also fail if a uk with the same column names exists? Can we also do:

```suggestion
  if (!is_empty(def$pks[[i]]$column) && identical(def$pks[[i]]$column[[1]], column)) {
    abort_no_uk_if_pk(table, column)
  }


  def$uks[[i]] <- vctrs::vec_rbind(
    def$uks[[i]],
    new_uk(list(column))
  )
```

Reply:
Good point, failing if a matching UK already exists makes a lot of sense. Need to adapt the error message then.
Thanks for your suggestion for checking for an existing, matching PK, looks good to me.

Reply:
Are we changing this here or in a future PR?